shell/platform/linux/fl_standard_message_codec.cc

```diff
@@ -0,0 +1,525 @@
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_standard_message_codec.h"
+#include "flutter/shell/platform/linux/fl_standard_message_codec_private.h"
+
+#include <gmodule.h>
+
+// See lib/src/services/message_codecs.dart in Flutter source for description of
+// encoding
+
+// Type values
+static constexpr int kValueNull = 0;
+static constexpr int kValueTrue = 1;
+static constexpr int kValueFalse = 2;
+static constexpr int kValueInt32 = 3;
+static constexpr int kValueInt64 = 4;
+static constexpr int kValueFloat64 = 6;
+static constexpr int kValueString = 7;
+static constexpr int kValueUint8List = 8;
+static constexpr int kValueInt32List = 9;
+static constexpr int kValueInt64List = 10;
+static constexpr int kValueFloat64List = 11;
+static constexpr int kValueList = 12;
+static constexpr int kValueMap = 13;
+
+// TODO(robert-ancell) Add support for large integer type (kValueLargeInt = 5)
+
+struct _FlStandardMessageCodec {
+  FlMessageCodec parent_instance;
+};
+
+G_DEFINE_TYPE(FlStandardMessageCodec,
+              fl_standard_message_codec,
+              fl_message_codec_get_type())
+
+// Functions to write standard C number types
+
+static void write_uint8(GByteArray* buffer, uint8_t value) {
+  g_byte_array_append(buffer, &value, sizeof(uint8_t));
+}
+
+static void write_uint16(GByteArray* buffer, uint16_t value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(uint16_t));
+}
+
+static void write_uint32(GByteArray* buffer, uint32_t value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(uint32_t));
+}
+
+static void write_int32(GByteArray* buffer, int32_t value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(int32_t));
+}
+
+static void write_int64(GByteArray* buffer, int64_t value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(int64_t));
+}
+
+static void write_float64(GByteArray* buffer, double value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(double));
+}
+
+// Checks there is enough data in the buffer to be read
+static gboolean check_size(GBytes* buffer,
+                           size_t* offset,
+                           size_t required,
+                           GError** error) {
+  if (*offset + required > g_bytes_get_size(buffer)) {
+    g_set_error(error, FL_MESSAGE_CODEC_ERROR,
+                FL_MESSAGE_CODEC_ERROR_OUT_OF_DATA, "Unexpected end of data");
+    return FALSE;
+  }
+  return TRUE;
+}
+
+// Gets a pointer to the current read location in the buffer
+static const uint8_t* get_data(GBytes* buffer, size_t* offset) {
+  return static_cast<const uint8_t*>(g_bytes_get_data(buffer, nullptr)) +
+         *offset;
+}
+
+// Functions to read FlValues in Standard encoding
+
+static gboolean read_uint8(GBytes* buffer,
```

Comment:
These should all have declaration comments, even though they will be pretty boilerplate-y, since there are non-obvious things like how errors are returned.